File: tests/r_parity/35_panel.py
"""StatsPAI panel FE / RE / Hausman parity (Python side) -- Module 35.

Generates a deterministic balanced panel with unit-level fixed
effects and runs sp.panel under both FE and RE specifications, plus
the Hausman test. The companion 35_panel.R uses plm::plm + plm::phtest.

Tolerance: rel < 1e-3 on the FE/RE coefficients and the plm-style
Hausman statistic.
"""
from __future__ import annotations

import numpy as np
import pandas as pd
import statspai as sp

from _common import PARITY_SEED, ParityRecord, dump_csv, write_results


MODULE = "35_panel"


def make_data(N: int = 100, T: int = 8, seed: int = PARITY_SEED) -> pd.DataFrame:
    rng = np.random.default_rng(seed)
    unit = np.repeat(np.arange(N), T)
    year = np.tile(np.arange(T), N)
    n = N * T
    x1 = rng.normal(size=n)
    x2 = rng.normal(size=n)
    unit_fe = rng.normal(0, 1, N)
    y = 1.0 + 0.5 * x1 - 0.3 * x2 + unit_fe[unit] + rng.normal(0, 0.5, n)
    return pd.DataFrame({"y": y, "x1": x1, "x2": x2,
                         "unit": unit, "year": year})


def main() -> None:
    df = make_data()
    dump_csv(df, MODULE)

    rows: list[ParityRecord] = []

    # FE
    fit_fe = sp.panel(data=df, formula="y ~ x1 + x2",
                      entity="unit", time="year", method="fe")
    for name in ["x1", "x2"]:
        rows.append(ParityRecord(
            module=MODULE, side="py", statistic=f"fe_beta_{name}",
            estimate=float(fit_fe.params[name]),
            se=float(fit_fe.std_errors[name]),
            n=int(len(df))))

    # RE
    fit_re = sp.panel(data=df, formula="y ~ x1 + x2",
                      entity="unit", time="year", method="re")
    for name in ["x1", "x2"]:
        rows.append(ParityRecord(
            module=MODULE, side="py", statistic=f"re_beta_{name}",
            estimate=float(fit_re.params[name]),
            se=float(fit_re.std_errors[name]),
            n=int(len(df))))

    # Hausman test (FE vs RE)
    haus = fit_fe.hausman_test()
    rows.append(ParityRecord(
        module=MODULE, side="py", statistic="hausman_chi2",
        estimate=float(haus["statistic"]),
        n=int(len(df))))
    rows.append(ParityRecord(
        module=MODULE, side="py", statistic="hausman_pvalue",
        estimate=float(haus["pvalue"]),
        n=int(len(df))))

    write_results(
        MODULE, "py", rows,
        extra={
            "N": 100, "T": 8,
            "hausman_parity_note": (
                "FE and RE coefficients match plm::plm at rel < 1e-15. "
                "The Hausman chi-squared statistic uses the same "
                "linearmodels/plm unadjusted FE-RE covariance as "
                "plm::phtest."
            ),
        },
    )


if __name__ == "__main__":
    main()

File: tests/r_parity/35_panel.R
# StatsPAI panel FE / RE / Hausman parity (R side) -- Module 35.
#
# Reads data/35_panel.csv and runs plm::plm + plm::phtest.
# Tolerance: rel < 1e-3 on FE coefficients.

.args <- commandArgs(trailingOnly = FALSE)
.file_arg <- grep("^--file=", .args, value = TRUE)
.script_dir <- if (length(.file_arg) > 0) {
  dirname(normalizePath(sub("^--file=", "", .file_arg[1])))
} else {
  getwd()
}
source(file.path(.script_dir, "_common.R"))

suppressPackageStartupMessages({
  library(plm)
})

MODULE <- "35_panel"

df <- read_csv_strict(MODULE)
pdf_obj <- plm::pdata.frame(df, index = c("unit", "year"))

fit_fe <- plm::plm(y ~ x1 + x2, data = pdf_obj, model = "within")
fit_re <- plm::plm(y ~ x1 + x2, data = pdf_obj, model = "random")
haus   <- plm::phtest(fit_fe, fit_re)

co_fe <- coef(fit_fe)
se_fe <- sqrt(diag(vcov(fit_fe)))
co_re <- coef(fit_re)
se_re <- sqrt(diag(vcov(fit_re)))

rows <- list(
  parity_row(MODULE, "fe_beta_x1",
             estimate = unname(co_fe["x1"]),
             se = unname(se_fe["x1"]), n = nrow(df)),
  parity_row(MODULE, "fe_beta_x2",
             estimate = unname(co_fe["x2"]),
             se = unname(se_fe["x2"]), n = nrow(df)),
  parity_row(MODULE, "re_beta_x1",
             estimate = unname(co_re["x1"]),
             se = unname(se_re["x1"]), n = nrow(df)),
  parity_row(MODULE, "re_beta_x2",
             estimate = unname(co_re["x2"]),
             se = unname(se_re["x2"]), n = nrow(df)),
  parity_row(MODULE, "hausman_chi2",
             estimate = unname(haus$statistic), n = nrow(df)),
  parity_row(MODULE, "hausman_pvalue",
             estimate = unname(haus$p.value), n = nrow(df))
)

write_results(MODULE, rows, extra = list(method_fe = "within",
                                          method_re = "random"))
